roachtest: add and fix acceptance/replicagc-changed-peers

tbg opened the pull request:
This fixes the failure of the acceptance test (see below)
since it will wake up dormant replicas, at which point they'll
become candidates and are checked for replicaGC aggressively.

This change also means that as a node restarts, all replicas are woken
up (once) over an interval of 10 minutes (default scan interval) when a
node restarts. This is a kind of "slow unquiescence storm" that should
not impact cluster performance significantly. For example, at 50k
ranges, we'd wake up ~83 ranges per second over the first ten minutes
after restarting a node (virtually all of these ranges are expected to
immediately quiesce again).

It's still unfortunate that this is sort of a side effect in the base
queue and driven by the scanner. Perhaps we could add a goroutine at
node start time that explicitly performs this task, and generates
appropriate output to indicate what's happening.

For the test:

Shut down a node, move all of its replicas to a completely disjoint set
of stores, and restart the node. Ensure that it replicaGCs the remaining
replicas reasonably fast (i.e. within minutes).

Note that #32843 would cause failures in this test as well, but this
problem has not been observed so far and are unlikely to occur in this
test since it leaves no freedom for replica placement at any time.

Fixes #25131.
Fixes #24067.

[1]: https://github.com/cockroachdb/cockroach/blob/cf617ba86cae5e6a8b7f977ebb49ea6f0afe6a95/pkg/storage/replica_gc_queue.go#L144-L155

Release note (bug fix): CockroachDB will no longer report erroneous under-replicated ranges corresponding to replicas that were waiting to be deleted.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32845)
<!-- Reviewable:end -->

tbg committed:
storage: wake up dormant ranges before replicaGC

This fixes the failure of the acceptance test in the next commit
since it will wake up dormant replicas, at which point they'll
become candidates and are checked for replicaGC aggressively.

This change also means that as a node restarts, all replicas are woken
up (once) over an interval of 10 minutes (default scan interval) when a
node restarts. This is a kind of "slow unquiescence storm" that should
not impact cluster performance significantly. For example, at 50k
ranges, we'd wake up ~83 ranges per second over the first ten minutes
after restarting a node (virtually all of these ranges are expected to
immediately quiesce again).

It's still unfortunate that this is sort of a side effect in the base
queue and driven by the scanner. Perhaps we could add a goroutine at
node start time that explicitly performs this task, and generates
appropriate output to indicate what's happening.

Fixes #25131.
Fixes #24067.

Release note (bug fix): CockroachDB will no longer report erroneous
under-replicated ranges corresponding to replicas that were waiting
to be deleted.

tbg commented:
I added a comment that lets the scanner wake up dormant replicas as it adds them to the replica GC queue, after which the test passes. I still need to figure out why I needed to restart n4-n6 in the test to get it to fail originally, and I need to fix the `--join` flag. But ready for a look!

tbg commented:
@petermattis friendly ping

petermattis commented:
Dropped off my radar. I’ll get to this tomorrow.

petermattis reviewed (approved):
I think the commit message math is off. 50k replicas over 10m will wake up at a rate of 83 per second, not per minute. While you're in the there, the `This change also means as a node restarts` also redundantly ends with `when a node restarts`.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice test!

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32845#-:-LTJjWFP4jgS2fYi1apL:b-yqtiyk)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/acceptance.go, line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32845#-LTJkTNO2EIFrMJmnIU8:-LTJkTNO2EIFrMJmnIU9:b-edavrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/b176f8cc1e6a1b24a3f6209835e16921c805ff26/pkg/cmd/roachtest/acceptance.go#L55)):*
> ```Go
> 		},
> 		6: {
> 			{"replicagc-changed-peers", runReplicaGCChangedPeers},
> ```

How long does this test make? I'm wondering if it needs to be an acceptance test. Note that the acceptance tests are run on every CI. 
___
*[pkg/cmd/roachtest/decommission.go, line 739 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32845#-LTJkfC709p8hSB0DFni:-LTJkfC709p8hSB0DFnj:b-axynv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b176f8cc1e6a1b24a3f6209835e16921c805ff26/pkg/cmd/roachtest/decommission.go#L739)):*
> ```Go
> 
> 	// Kill the third node so it won't know that all of its replicas are moved
> 	// elsewhere. (We don't use the first because that's what roachtest will
> ```

`s/roachtest/roachprod/g`
___
*[pkg/cmd/roachtest/decommission.go, line 783 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32845#-LTJksDP-XO420q0BYQa:-LTJksDP-XO420q0BYQb:b-5tci3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/b176f8cc1e6a1b24a3f6209835e16921c805ff26/pkg/cmd/roachtest/decommission.go#L783)):*
> ```Go
> 	// won't contact the replicas of n3. Perhaps the real reason is the loss
> 	// of Gossip state. Maybe n3 catches wind of its having been decommissioned
> 	// that way?
> ```

Did you ever track down the answer here?
___
*[pkg/storage/queue.go, line 211 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32845#-LTJjxUp1G7HFh9m3ZIi:-LTJjxUp1G7HFh9m3ZIj:bw6n8oh) ([raw file](https://github.com/cockroachdb/cockroach/blob/42fe5c0575654c14aa616210c0f03e18aa12153f/pkg/storage/queue.go#L211)):*
> ```Go
> 	// needsNotDormant controls whether the Raft group will be initialized (if
> 	// not already initialized) when deciding whether to process this replica.
> 	needsNotDormant bool
> ```

That phrasing here feels too negative, but I'm not sure what the opposite of dormant is in this context. Perhaps `needsRaftInitialized`. 


<!-- Sent from Reviewable.io -->

tbg committed:
roachtest: add acceptance/replicagc-changed-peers

Shut down a node, move all of its replicas to a completely disjoint set
of stores, and restart the node. Ensure that it replicaGCs the remaining
replicas reasonably fast (i.e. within minutes).

Note that #32843 would cause failures in this test as well, but this
problem has not been observed so far and are unlikely to occur in this
test since it leaves no freedom for replica placement at any time.

Touches #25131.
Touches #24067.

Release note: None

tbg reviewed (commented):
> I think the commit message math is off

Thanks for catching that, fixed.

I figured out why I needed to restart the remainder of the cluster. When n3 restarts and n4 and n5 are still running, n3 learns that n1 and n2 are decommissioning.
As a consequence, n3's replicate queue's shouldQueue advises repairs for all of its ranges (which have replicas on n1 and n2) and that acquires a lease, which inits the Raft group and then it ends up as candidate and gets processed for replicaGC. Updated the test to recommission the other nodes, and also added two flavors to make sure it works with the restart and without.

TFTR!

bors r=petermattis

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32845#-:-LTa_4LM95j-Jla7Si33:bll0d9z)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/acceptance.go, line 55 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32845#-LTJkTNO2EIFrMJmnIU8:-LTacHtqBHfACpY715xw:b8z0wrq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b176f8cc1e6a1b24a3f6209835e16921c805ff26/pkg/cmd/roachtest/acceptance.go#L55)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

How long does this test make? I'm wondering if it needs to be an acceptance test. Note that the acceptance tests are run on every CI. 
</blockquote></details>

Yeah, I don't think it needs to be an acceptance test. Runs in about 2 minutes or less.
___
*[pkg/cmd/roachtest/decommission.go, line 739 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32845#-LTJkfC709p8hSB0DFni:-LTb5d4d0Hjzf92wb4Zy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b176f8cc1e6a1b24a3f6209835e16921c805ff26/pkg/cmd/roachtest/decommission.go#L739)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/roachtest/roachprod/g`
</blockquote></details>

Done.
___
*[pkg/cmd/roachtest/decommission.go, line 783 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32845#-LTJksDP-XO420q0BYQa:-LTb5dT96HfEU_0GThJ3:b-14m9ye) ([raw file](https://github.com/cockroachdb/cockroach/blob/b176f8cc1e6a1b24a3f6209835e16921c805ff26/pkg/cmd/roachtest/decommission.go#L783)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Did you ever track down the answer here?
</blockquote></details>

See comment on the main thread.
___
*[pkg/storage/queue.go, line 211 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32845#-LTJjxUp1G7HFh9m3ZIi:-LTb6CNN6tGTHlfyRZtp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/42fe5c0575654c14aa616210c0f03e18aa12153f/pkg/storage/queue.go#L211)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

That phrasing here feels too negative, but I'm not sure what the opposite of dormant is in this context. Perhaps `needsRaftInitialized`. 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
